src/scheduler/check.cpp

```diff
@@ -1571,26 +1571,44 @@ check_normal_node_path(status *policy, server_info *sinfo, queue_info *qinfo, re
 			if (qinfo->has_nodes)
 				ninfo_arr = qinfo->nodes;
 			else if (!(sinfo->svr_to_psets.empty())) {	/* Multi-server psets */
+				int i = 0;
+				int owner_set = -1;
+
 				/* Find the pset of the server which owns the job */
 				for (auto &spset: sinfo->svr_to_psets) {
 					if (spset.svr_inst_id == resresv->job->svr_inst_id) {
 						msvr_pset[0] = spset.np;
-						if (!resresv->job->is_array && spec->total_chunks == 1) {
-							/* Allow job to be placed on nodes of other servers as well
-							 * For now, restricting job arrays and multi-chunk jobs
-							 * to nodes of just the local server
-							 */
-							msvr_pset[1] = sinfo->allpart;
-							msvr_pset[2] = NULL;
+						owner_set = i;
+						break;
+					}
+					i++;
+				}
+				if (owner_set != -1) {
+					if (resresv->job->is_array)	/* Restrict job arrays to owner server */
+						msvr_pset[1] = NULL;
+					else if (spec->total_chunks == 1) {	/* Single chunk jobs */
+						msvr_pset[1] = sinfo->allpart;	/* If owner's nodes don't work, use all */
+						msvr_pset[2] = NULL;
+					} else {	/* Potential multi-node jobs */
+						/* If owner nodes don't work, try to first fit fully on a one server */
+						i = 0;
+						int j = 1;	/* index of msvr_pset */
+						for (auto &spset: sinfo->svr_to_psets) {
+							if (i != owner_set)
+								msvr_pset[j++] = spset.np;
+							i++;
+						}
+						if (resresv->is_resv) {
```

Comment:
We will enter into this block if it is a job but not a reservation. This is due to the condition at the line 1569. 
if (resresv->is_job) {
  .....
}
I think we can either remove this if condition or change it to the following which also takes care or reservations.
if (resresv->is_job || resresv->resv)

Reply:
Thanks for catching this! if (resresv->is_job || resresv->is_resv) is as good as removing the if condition, the resresv can only be a job or a reservation. Let me think about it and see what I can do